### PR TITLE
add support for setting the resolution of ds18b20 sensors

### PR DESCRIPTION
**Pull request by KoffeinKaio**
Hey,

I'm slowly getting in the pilight code and thought of something easy to get started.

Since Linux Kernel version ~4.7, the w1-therm driver supports setting the resolution of the ds18b20 sensors by writing to the dev file.

This change adds a 'resolution' setting to the protocol which allows the configuration for 9/10/11/12 bit resolution

### CODE REVIEW

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 1]
```diff
@@ -96,8 +96,13 @@ static void *ds18b20Parse(void *param) {
 
 	if(json_find_number(json, "poll-interval", &itmp) == 0)
 		interval = (int)round(itmp);
+
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: Extra blank line

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 2]
```diff
@@ -96,8 +96,13 @@ static void *ds18b20Parse(void *param) {
...
 	json_find_number(json, "temperature-offset", &temp_offset);
 
+	if(json_find_number(json, "resolution", &rtemp) == 0)
+		resolution = (int)round(rtemp);
+
+
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: More extra blank lines (please remove)

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 3]
```diff
@@ -133,6 +138,17 @@ static void *ds18b20Parse(void *param) {
...
+                                //try to set resolution here - set it temporary to not wear out the EEEPROM
+                                if ((rfd = fopen(ds18b20_w1slave, "w"))) {
+                                    logprintf(LOG_DEBUG, "setting resolution of %s to: %d", ds18b20_w1slave, resolution);
+                                    if (fprintf(ffd, "%d", resolution) < 0 )
+                                        logprintf(LOG_ERR, "cannot set resolution of %s", ds18b20_w1slave);
+
+                                    fclose(ffd);
+                                } else {
+                                    logprintf(LOG_ERR, "opening %s to set resolution failed!", ds18b20_w1slave);
+                                }
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: Use tabs to indent the code in this section

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 4]
```diff
@@ -133,6 +138,17 @@ static void *ds18b20Parse(void *param) {
 								}
 								memset(content, '\0', bytes+1);
 
+                                //try to set resolution here - set it temporary to not wear out the EEEPROM
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: `// Try to set resolution here - set it temporary to not wear out the EEPROM`

(grammar/spelling)

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 5]
```diff
@@ -133,6 +138,17 @@ static void *ds18b20Parse(void *param) {
...
+                                if ((rfd = fopen(ds18b20_w1slave, "w"))) {
+                                    logprintf(LOG_DEBUG, "setting resolution of %s to: %d", ds18b20_w1slave, resolution);
+                                    if (fprintf(ffd, "%d", resolution) < 0 )
+                                        logprintf(LOG_ERR, "cannot set resolution of %s", ds18b20_w1slave);
+
+                                    fclose(ffd);
+                                } else {
+                                    logprintf(LOG_ERR, "opening %s to set resolution failed!", ds18b20_w1slave);
+                                }
+
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: spurious blank line

**Comment by pilino1234** [comment]
Looks pretty good!

How will this behave on devices running a kernel < 4.7?

**Comment by pilino1234** [comment]
Also, can you document the wiki page for this protocol: https://wiki.pilight.org/doku.php/ds18b20_v7_0?

You can create a new page based on the link above, and expand [this table](https://wiki.pilight.org/doku.php/protocols#gpio_connected_sensors) to include a new column labelled "7.0 dev" and a link to your new page?

File: libs/pilight/protocols/GPIO/ds18b20.c (hunk 5 above)
Comment (KoffeinKaio) [reply libs/pilight/protocols/GPIO/ds18b20.c]: You sure about that one?
Removing this blank line  sets the next if right below - kinda destroys readability

**Comment by KoffeinKaio** [comment]
On kernels below <4.7 it will probably throw an error that the device is write protected or something
Maybe I should check the kernel version or fail silently

File: libs/pilight/protocols/GPIO/ds18b20.c (hunk 5 above)
Comment (pilino1234) [reply libs/pilight/protocols/GPIO/ds18b20.c]: Ah, never mind then :)

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 6]
```diff
@@ -138,16 +135,15 @@ static void *ds18b20Parse(void *param) {
...
 								memset(content, '\0', bytes+1);
 
-                                //try to set resolution here - set it temporary to not wear out the EEEPROM
-                                if ((rfd = fopen(ds18b20_w1slave, "w"))) {
-                                    logprintf(LOG_DEBUG, "setting resolution of %s to: %d", ds18b20_w1slave, resolution);
-                                    if (fprintf(ffd, "%d", resolution) < 0 )
-                                        logprintf(LOG_ERR, "cannot set resolution of %s", ds18b20_w1slave);
+								if ((rfd = fopen(ds18b20_w1slave, "w"))) {
+									logprintf(LOG_DEBUG, "setting resolution of %s to: %d", ds18b20_w1slave, resolution);
+									if (fprintf(ffd, "%d", resolution) < 0 )
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: Add braces for one-line if statement bodies too

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 7]
```diff
@@ -135,14 +139,21 @@ static void *ds18b20Parse(void *param) {
...
+										if ((rfd = fopen(ds18b20_w1slave, "w"))) {
+											logprintf(LOG_DEBUG, "setting resolution of %s to: %d", ds18b20_w1slave, resolution);
+											if (fprintf(rfd, "%d", resolution) < 0 ) {
+												logprintf(LOG_ERR, "cannot set resolution of %s", ds18b20_w1slave);
+											}
+											fclose(rfd);
+										} else {
+											logprintf(LOG_ERR, "opening %s to set resolution failed!", ds18b20_w1slave);
+										}
+									}
```
Comment (pilino1234) [inline libs/pilight/protocols/GPIO/ds18b20.c]: Can there be an else here, warning the user that their settings isn't going to work?

Comment (KoffeinKaio) [reply libs/pilight/protocols/GPIO/ds18b20.c]: sure

**Comment by pilino1234** [comment]
Looks good! Just needs a squash and the aforementioned documentation on the pilight wiki :smile:

File: libs/pilight/protocols/GPIO/ds18b20.c (hunk 6 above)
Comment (CurlyMoo) [reply libs/pilight/protocols/GPIO/ds18b20.c]: The spaces between the `if` and the hooks should also be removed.

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 8]
```diff
@@ -133,6 +139,27 @@ static void *ds18b20Parse(void *param) {
 								}
 								memset(content, '\0', bytes+1);
 
+								if (uname(&unameData) != -1) {
+									char **varr = NULL;
+									int toks = explode(unameData.release, ".", &varr);
```
Comment (CurlyMoo) [inline libs/pilight/protocols/GPIO/ds18b20.c]: Have you checked for memory leaks with valgrind? You aren't freeing the `&varr` anywhere?

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 9]
```diff
@@ -133,6 +139,27 @@ static void *ds18b20Parse(void *param) {
...
+											if (fprintf(rfd, "%d", resolution) < 0 ) {
+												logprintf(LOG_ERR, "cannot set resolution of %s", ds18b20_w1slave);
+											}
+											fclose(rfd);
+										} else {
+											logprintf(LOG_ERR, "opening %s to set resolution failed!", ds18b20_w1slave);
+										}
+									} else {
+										if (rtmp != 0) {
+											logprintf(LOG_ERR, "kernel version >=4.7 is required for setting resolution on ds18b20 sensors. ");
```
Comment (CurlyMoo) [inline libs/pilight/protocols/GPIO/ds18b20.c]: No space between the operator and number and too many spaces afterwards.

File: libs/pilight/protocols/GPIO/ds18b20.c [hunk 10]
```diff
@@ -133,6 +139,27 @@ static void *ds18b20Parse(void *param) {
...
+
+									if (toks > 2 && strtol(varr[0], NULL, 10) >= 4 && strtol(varr[1], NULL, 10) >= 7) {
+										if ((rfd = fopen(ds18b20_w1slave, "w"))) {
+											logprintf(LOG_DEBUG, "setting resolution of %s to: %d", ds18b20_w1slave, resolution);
+											if (fprintf(rfd, "%d", resolution) < 0 ) {
+												logprintf(LOG_ERR, "cannot set resolution of %s", ds18b20_w1slave);
+											}
+											fclose(rfd);
+										} else {
+											logprintf(LOG_ERR, "opening %s to set resolution failed!", ds18b20_w1slave);
```
Comment (CurlyMoo) [inline libs/pilight/protocols/GPIO/ds18b20.c]: Please don't use exclamation marks.

**Comment by CurlyMoo** [comment]
In general, i would like to see a PR to the rewrite branch first that is fully unittested. Then we can backport everything to development.

**Comment by KoffeinKaio** [comment]
I'll do a new pr to rewrite tomorrow

**Comment by CurlyMoo** [comment]
Please don't just close PR's.

**Pull request closed without merging**